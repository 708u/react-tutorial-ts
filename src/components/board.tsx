import React, { FC, MouseEvent } from 'react';
import { v4 as uuidV4 } from 'uuid';
import Square, { mark } from 'components/square';
import styled from 'styled-components';
import 'index.css';

export type board = mark[];

type Props = {
  squares: board;
  onClick: (n: number) => (e: MouseEvent<HTMLButtonElement>) => void;
};

type LocationIndex = '1-1' | '1-2' | '1-3' | '2-1' | '2-2' | '2-3' | '3-1' | '3-2' | '3-3';

const LocationMap: Map<
  LocationIndex,
  {
    row: number;
    col: number;
  }
> = new Map([
  ['1-1', { row: 1, col: 1 }],
  ['1-2', { row: 1, col: 2 }],
  ['1-3', { row: 1, col: 3 }],
  ['2-1', { row: 2, col: 1 }],
  ['2-2', { row: 2, col: 2 }],
  ['2-3', { row: 2, col: 3 }],
  ['3-1', { row: 3, col: 1 }],
  ['3-2', { row: 3, col: 2 }],
  ['3-3', { row: 3, col: 3 }],
]);

const Board: FC<Props> = (props: Props) => {
  const BoardRaw = styled.div`
    ::after {
      clear: both;
      content: '';
      display: table;
    }
  `;

  const { squares, onClick } = props;

  const boardNumbers: {
    location: LocationIndex;
    id: number;
  }[][] = [
    [
      {
        location: '1-1',
        id: 0,
      },
      {
        location: '1-2',
        id: 1,
      },
      {
        location: '1-3',
        id: 2,
      },
    ],
    [
      {
        location: '2-1',
        id: 3,
      },
      {
        location: '2-2',
        id: 4,
      },
      {
        location: '2-3',
        id: 5,
      },
    ],
    [
      {
        location: '3-1',
        id: 6,
      },
      {
        location: '3-2',
        id: 7,
      },
      {
        location: '3-3',
        id: 8,
      },
    ],
  ];

  return (
    <div>
      {boardNumbers.map((row) => (
        <BoardRaw key={`board-row-${uuidV4()}`}>
          {row.map((value) => (
            <Square
              key={`square-btn-${uuidV4()}`}
              name={value.location}
              testId={value.id}
              value={squares[value.id]}
              onClick={onClick(value.id)}
            />
          ))}
        </BoardRaw>
      ))}
    </div>
  );
};

export type { LocationIndex };

export { LocationMap };

export default Board;
