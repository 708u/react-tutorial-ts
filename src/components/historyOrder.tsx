import React, { FC } from 'react';
import styled from 'styled-components';

const HistoryOrder: FC = () => {
  const Body = styled.div`
    margin-top: 10px;
    margin-left: 30px;
  `;

  return (
    <Body>
      <button type="button">Toggle Asc/Desc</button>
    </Body>
  );
};

export default HistoryOrder;
