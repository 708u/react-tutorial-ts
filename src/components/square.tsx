import React, { FC, MouseEvent } from 'react';
import styled from 'styled-components';

export type mark = 'O' | 'X' | null;

export const SquareBtn = styled.button`
  background: #fff;
  border: 1px solid #999;
  float: left;
  font-size: 24px;
  font-weight: bold;
  line-height: 34px;
  height: 34px;
  margin-right: -1px;
  margin-top: -1px;
  padding: 0;
  text-align: center;
  width: 34px;
`;

type Props = {
  onClick: (e: MouseEvent<HTMLButtonElement>) => void;
  value: mark;
  name: string;
  testId?: number;
};

const Square: FC<Props> = (props) => {
  const { onClick, name, value, testId } = props;
  return (
    <SquareBtn name={name} data-testid={`btn-${testId}`} onClick={onClick}>
      {value}
    </SquareBtn>
  );
};

export default Square;
