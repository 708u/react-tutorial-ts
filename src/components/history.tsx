import React, { FC } from 'react';
import { v4 as uuidV4 } from 'uuid';
import { board } from 'components/board';
import styled from 'styled-components';

export type Position = {
  col: number;
  row: number;
};

export type squares = {
  squares: board;
  position?: Position;
};

export type History = squares[];

type Props = {
  history: History;
  current: number;
  onClick: (move: number) => () => void;
};

const History: FC<Props> = (props) => {
  const HistoryList = styled.ol`
    padding-left: 30px;
  `;

  const { history, onClick, current } = props;

  return (
    <HistoryList>
      {history.map((h, move) => (
        <li key={`history-list-${uuidV4()}`}>
          <button type="button" onClick={onClick(move)} className={current === move ? 'btn-bold' : ''}>
            {move ? `Go to move #${move}. [row: ${h.position?.row}, col: ${h.position?.col}]` : 'Go to game start'}
          </button>
        </li>
      ))}
    </HistoryList>
  );
};

export default History;
