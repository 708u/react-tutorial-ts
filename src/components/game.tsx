import React, { useState, MouseEvent, FC } from 'react';
import Board, { board, LocationMap, LocationIndex } from 'components/board';
import { mark } from 'components/square';
import History, { History as HistoryType } from 'components/history';
import styled from 'styled-components';
import 'index.css';

const Game: FC = () => {
  const View = styled.div`
    flex-direction: row;
    display: flex;
    font: 14px 'Century Gothic', Futura, sans-serif;
    margin: 20px;
  `;

  const GameInfo = styled.div`
    margin-left: 20px;
  `;

  const [history, setHistory] = useState<HistoryType>([
    {
      squares: Array(9).fill(null),
    },
  ]);
  const [stepNumber, setStepNumber] = useState(0);
  const [xIsNext, setXIsNext] = useState(true);
  const current = history[stepNumber];

  const gameIsOverPatterns = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ] as const;

  // FIXME: refactoring arg name;
  const calculateWinner = (bd: board): mark => {
    let winner: mark = null;
    gameIsOverPatterns.forEach((v, idx) => {
      const [a, b, c] = gameIsOverPatterns[idx];
      // X, null, X
      if (bd[a] && bd[a] === bd[b] && bd[a] === bd[c]) {
        winner = bd[a]; // 勝利したマークを返却する
      }
    });
    return winner;
  };

  const handleClick = (i: number) => (e: MouseEvent<HTMLButtonElement>) => {
    // すべての盤面履歴取得
    // history = [{hist1}, {hist2}]
    const newHistory = history.slice(0, stepNumber + 1);
    // 盤面全体を保持するobject取得
    // { squares: Array(9)}
    const newSquare = newHistory[newHistory.length - 1];
    // 現在の盤面を取得しつつ、インスタンスを新規に作成
    // ['', 'X', ...];
    const copiedSquares = newSquare.squares.slice();

    // クリックした点で勝利した人がいたらゲームを早期終了
    if (calculateWinner(copiedSquares) || copiedSquares[i]) {
      return;
    }
    // 盤面にマークをセット
    copiedSquares[i] = xIsNext ? 'X' : 'O';

    const position = e.currentTarget.name as LocationIndex;

    // state更新
    setHistory(
      newHistory.concat([
        {
          squares: copiedSquares,
          position: LocationMap.get(position),
        },
      ])
    );
    setStepNumber(newHistory.length);
    setXIsNext(!xIsNext);
  };

  const jumpTo = (step: number) => () => {
    setStepNumber(step);
    setXIsNext(step % 2 === 0);
  };

  const getStatus = () => {
    const winner = calculateWinner(current.squares);
    return winner ? `Winner: ${winner}` : `Next player: ${xIsNext ? 'X' : 'O'}`;
  };

  return (
    <View>
      <Board squares={current.squares} onClick={handleClick} />
      <GameInfo>
        <GameInfo>{getStatus()}</GameInfo>
        <History history={history} current={stepNumber} onClick={jumpTo} />
      </GameInfo>
    </View>
  );
};

export default Game;
