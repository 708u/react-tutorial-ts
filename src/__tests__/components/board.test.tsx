import renderer from 'react-test-renderer';
import React, { MouseEvent } from 'react';
import Board from 'components/board';
import { mark } from 'components/square';

describe('testing board components', () => {
  it('should be same as a previous snapshot.', () => {
    const squares: mark[] = ['O', 'X', 'O', 'X', 'O', 'X', 'O', 'X', 'O'];

    /* eslint-disable @typescript-eslint/no-unused-vars */
    const onClick = (i: number) => (e: MouseEvent<HTMLButtonElement>) => true;
    const component = renderer.create(<Board squares={squares} onClick={onClick} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
