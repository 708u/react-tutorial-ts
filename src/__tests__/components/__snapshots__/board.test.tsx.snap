// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing board components should be same as a previous snapshot. 1`] = `
<div>
  <div
    className="sc-gsTCUz kQeNae"
  >
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-0"
      name="1-1"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-1"
      name="1-2"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-2"
      name="1-3"
      onClick={[Function]}
    >
      O
    </button>
  </div>
  <div
    className="sc-gsTCUz kQeNae"
  >
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-3"
      name="2-1"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-4"
      name="2-2"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-5"
      name="2-3"
      onClick={[Function]}
    >
      X
    </button>
  </div>
  <div
    className="sc-gsTCUz kQeNae"
  >
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-6"
      name="3-1"
      onClick={[Function]}
    >
      O
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-7"
      name="3-2"
      onClick={[Function]}
    >
      X
    </button>
    <button
      className="sc-bdfBwQ iZlURE"
      data-testid="btn-8"
      name="3-3"
      onClick={[Function]}
    >
      O
    </button>
  </div>
</div>
`;
